closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -0,0 +1,1040 @@
+                        if (input.length != 3) {
+                            printStackTraceAndExit("Constraint format is incorrect: " + providerInfo);
+                        }
+
+                        String inType = input[0].trim();
+                        String inAlgorithm = input[1].trim();
+                        String inAttributes = input[2].trim();
+
+                        // Each attribute must includes 2 fields (key and value) or *.
+                        if (!isAsterisk(inAttributes)) {
+                            String[] attributeArray = inAttributes.split(":");
+                            for (String attribute : attributeArray) {
+                                String[] in = attribute.split("=", 2);
+                                if (in.length != 2) {
+                                    printStackTraceAndExit(
+                                            "Constraint attributes format is incorrect: " + providerInfo);
+                                }
+                            }
+                        }
+
+                        Constraint constraint = new Constraint(inType, inAlgorithm, inAttributes);
+
+                        if (debug != null) {
+                            debug.println("Loading constraints for security provider: " + providerName
+                                    + " with constraints type: " + constraint.type
+                                    + " algorithm: " + constraint.algorithm
+                                    + " attributes: " + constraint.attributes);
+                        }
+                        constraints[cNum] = constraint;
+                        cNum++;
+                    }
+                    providerConstraints.put(providerName, constraints);
+                    if (debug != null) {
+                        debug.println("Loaded constraints for security provider: " + providerName);
+                    }
+                } else {
+                    printStackTraceAndExit("Constraint format is incorrect: " + providerInfo);
+                }
+            }
+        }
+
+        /**
+         * Check if the Service is allowed in restricted security mode.
+         *
+         * @param service the Service to check
+         * @return true if the Service is allowed
+         */
+        boolean isRestrictedServiceAllowed(Service service) {
+            String providerName = service.getProvider().getName();
+
+            // Provider with argument, remove argument.
+            // e.g. SunPKCS11-NSS-FIPS, remove argument -NSS-FIPS.
+            int pos = providerName.indexOf('-');
+            providerName = (pos < 0) ? providerName : providerName.substring(0, pos);
+
+            Constraint[] constraints = providerConstraints.get(providerName);
+
+            if ((constraints == null) || (constraints.length == 0)) {
+                // Allow this provider with no constraints.
+                return true;
+            }
```

Comment:
This doesn't seem right. Services from a provider not mentioned in `java.security` should not be allowed.
Perhaps this should be:
```java
            if (constraints == null) {
                // Disallow unknown providers.
                return false;
            } else if (constraints.length == 0) {
                // Allow this provider with no constraints.
                return true;
            }
```